internal/controllers/selfhealing/slice.go

```diff
@@ -0,0 +1,188 @@
+package selfhealing
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+	"github.com/go-logr/logr"
+)
+
+// sliceController check if the resource slice is deleted but it is still present in the composition current synthesis status.
+// If yes, it will update the composition PendingResynthesis status to trigger re-synthesis process.
+type sliceController struct {
+	client                 client.Client
+	noCacheReader          client.Reader
+	selfHealingGracePeriod time.Duration
+}
+
+func NewSliceController(mgr ctrl.Manager, selfHealingGracePeriod time.Duration) error {
+	s := &sliceController{
+		client:                 mgr.GetClient(),
+		noCacheReader:          mgr.GetAPIReader(),
+		selfHealingGracePeriod: selfHealingGracePeriod,
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("selfHealingSliceController").
+		Watches(&apiv1.Composition{}, newCompositionHandler()).
+		Watches(&apiv1.ResourceSlice{}, newSliceHandler()).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "selfHealingSliceController")).
+		Complete(s)
+}
+
+func (s *sliceController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	comp := &apiv1.Composition{}
+	err := s.client.Get(ctx, req.NamespacedName, comp)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting composition: %w", err)
+	}
+
+	syn := &apiv1.Synthesizer{}
+	syn.Name = comp.Spec.Synthesizer.Name
+	err = s.client.Get(ctx, client.ObjectKeyFromObject(syn), syn)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("gettting synthesizer: %w", err))
+	}
+
+	logger = logger.WithValues("compositionGeneration", comp.Generation,
+		"compositionName", comp.Name,
+		"compositionNamespace", comp.Namespace,
+		"synthesisID", comp.Status.GetCurrentSynthesisUUID())
+
+	// Skip if the composition is not eligible for resynthesis, and check the synthesis result later
+	if notEligibleForResynthesis(comp) {
+		logger.V(1).Info("not eligible for resynthesis when checking the missing resource slice")
+		// Use default grace period
+		if comp.Status.CurrentSynthesis == nil ||
+			comp.Status.CurrentSynthesis.Synthesized == nil ||
+			(s.selfHealingGracePeriod-time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)) <= 0 {
+			return ctrl.Result{Requeue: true, RequeueAfter: s.selfHealingGracePeriod}, nil
+		}
+		return ctrl.Result{Requeue: true, RequeueAfter: s.selfHealingGracePeriod - time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)}, nil
```

Comment:
Is this code reachable? I think we need to move the timestamp/graceperiod comparison into  `notEligibleForResynthesis`

Reply:
The condition "(s.selfHealingGracePeriod-time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)) <= 0" is used to ensure we won't requeue request after a negative time, not for checking eligible to resynthesis.

I think the code is reachable. If comp.Status.CurrentSynthesis.Synthesized is existed and comp.Status.PendingResynthesis != nil, it will go into the condition "if notEligibleForResynthesis(comp) { ... }". And if time since last synthesized less than grace period e.g. (s.selfHealingGracePeriod-time.Since(comp.Status.CurrentSynthesis.Synthesized.Time)) > 0, it will use the remaining grace period as requeue after time.